src/scanner.rs

```diff
@@ -397,172 +420,281 @@ impl ScannerWorker {
+                            self.apply_character_filter(&peripheral, v, &mut passed).await;
+                        }
+                    }
+                }
 
-        if self.connection_needed {
-            if !peripheral.is_connected().await.unwrap_or(false) {
-                if self.connecting.lock().unwrap().insert(peripheral.id()) {
-                    log::debug!("Connecting to device {}", peripheral.address());
-
-                    // Connect in another thread, so we can keep filtering other devices meanwhile.
-                    let peripheral_clone = peripheral.clone();
-                    let connecting_map = self.connecting.clone();
-                    tokio::spawn(async move {
-                        if let Err(e) = peripheral_clone.connect().await {
-                            log::warn!(
-                                "Could not connect to {}: {:?}",
-                                peripheral_clone.address(),
-                                e
-                            );
-
-                            connecting_map
-                                .lock()
-                                .unwrap()
-                                .remove(&peripheral_clone.id());
-                        };
-                    });
+                if passed {
+                    self.matched.insert(peripheral_id.clone());
+                    self.result_count += 1;
+
+                    if let Err(e) = self.event_sender.send(
+                        DeviceEvent::Discovered(
+                            Device::new(self.session.adapter.clone(), peripheral))
+                    ) {
+                        log::warn!("error: {} when sending device", e);
+                    }
                 }
-                return;
-            } else if let Some(false) = self.passes_post_connect_filters(&peripheral).await {
-                self.skip_peripheral(&peripheral).await;
-                return;
+
+                log::debug!(
+                    "current matched: {}, current filtered: {}",
+                    self.matched.len(),
+                    self.filtered.len()
+                );
             }
-        } else if self.config.force_disconnect {
-            peripheral.disconnect().await.ok();
-        }
 
-        self.add_peripheral(peripheral).await;
+            self.filtered.insert(peripheral_id);
+        }
     }
 
-    async fn skip_peripheral(&mut self, peripheral: &Peripheral) {
-        self.filtered.insert(peripheral.id());
+    /// TODO validate
+    async fn apply_character_filter(&self, peripheral: &Peripheral, uuid: &Uuid, passed: &mut bool) {
+        if !peripheral.is_connected().await.unwrap_or(false) {
+            if self.connecting.lock().unwrap().insert(peripheral.id()) {
+                log::debug!("Connecting to device {}", peripheral.address());
+
+                // Connect in another thread, so we can keep filtering other devices meanwhile.
```

Comment:
This no longer connects in a separate thread, so this comment should be removed.

Reply:
Getting the characteristics requires connecting the device.

Reply:
Yes, I just meant that the old code did the connecting in a separate thread: `tokio::spawn(async move { ...`. This one doesn't, but may that is not a problem.

Reply:
It’s in your hands.